marshal/cluster.go

```diff
@@ -240,38 +240,76 @@ func (c *KafkaCluster) getClaimPartition(topicName string) int {
 	return int(uval % uint64(c.partitions))
 }
 
-// getPartitionState returns a topicState and possibly creates it and the partition state within
-// the State.
-func (c *KafkaCluster) getPartitionState(groupID, topicName string, partID int) *topicState {
+// getGroupState returns the map of topics to topicState objects for a group.
+func (c *KafkaCluster) getGroupState(groupID string) map[string]*topicState {
+	// Read lock check
+	c.lock.RLock()
+	if group, ok := c.groups[groupID]; ok {
+		c.lock.RUnlock()
+		return group
+	}
+	c.lock.RUnlock()
+
+	// Failed, write lock check and possible create
 	c.lock.Lock()
 	defer c.lock.Unlock()
 
-	group, ok := c.groups[groupID]
-	if !ok {
-		group = make(map[string]*topicState)
-		c.groups[groupID] = group
+	if group, ok := c.groups[groupID]; ok {
+		return group
 	}
+	c.groups[groupID] = make(map[string]*topicState)
+	return c.groups[groupID]
+}
 
-	topic, ok := group[topicName]
-	if !ok {
-		topic = &topicState{
-			claimPartition: c.getClaimPartition(topicName),
-			partitions:     make([]PartitionClaim, partID+1),
-		}
-		group[topicName] = topic
+// getTopicState returns a topicState for a given topic.
+func (c *KafkaCluster) getTopicState(groupID, topicName string) *topicState {
+	group := c.getGroupState(groupID)
+
+	// Read lock check
+	c.lock.RLock()
+	if topic, ok := group[topicName]; ok {
+		c.lock.RUnlock()
+		return topic
 	}
+	c.lock.RUnlock()
 
-	// Take the topic lock if we can
+	// Write lock check and possible create
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	if topic, ok := group[topicName]; ok {
+		return topic
+	}
+	group[topicName] = &topicState{
+		claimPartition: c.getClaimPartition(topicName),
+		partitions:     nil,
+	}
+	return group[topicName]
+}
+
+// getPartitionState returns a topicState and possibly creates it and the partition state within
+// the State.
+func (c *KafkaCluster) getPartitionState(groupID, topicName string, partID int) *topicState {
+	// Get topic and lock it so we can update it if needed
+	topic := c.getTopicState(groupID, topicName)
+
+	// UNLOCK NOT DEFERRED. This is so we can upgrade the lock.
```

Comment:
99% of the time when I hear someone say "lock upgrade" it is meant to be atomic / no one else can get the lock in between. This function just does an unlock followed by a lock which isn't an upgrade. If `if len(topic.partitions) > partID` is in a helper function it can do the regular defer on the read lock, and it won't change the locking semantics.